state/src/storage/tx/matching_pools.rs

```diff
@@ -0,0 +1,206 @@
+//! Helpers for accessing information about matching pools in the database
+
+use common::types::wallet::OrderIdentifier;
+use libmdbx::{TransactionKind, RW};
+
+use crate::{storage::error::StorageError, POOL_TABLE};
+
+use super::StateTxn;
+
+// -------------
+// | Constants |
+// -------------
+
+/// The name of the global matching pool
+pub const GLOBAL_MATCHING_POOL: &str = "global";
+
+/// The prefix of all matching pool keys
+pub const POOL_KEY_PREFIX: &str = "matching-pool/";
+
+/// The error message used when trying to create a pool that already exists
+const POOL_EXISTS_ERR: &str = "pool already exists";
+/// The error message used when allocating an order to a nonexistent pool
+const POOL_DOES_NOT_EXIST_ERR: &str = "pool does not exist";
+/// The error message used when a non-empty pool is attempted to be destroyed
+const POOL_NOT_EMPTY_ERR: &str = "pool not empty";
+
+// ---------------
+// | Key Helpers |
+// ---------------
+
+/// The key for the set of all matching pools
+pub fn all_pools_key() -> String {
+    "all-pools".to_string()
+}
+
+/// The key for the given order's matching pool
+pub fn matching_pool_key(order_id: &OrderIdentifier) -> String {
+    format!("{}{}", POOL_KEY_PREFIX, order_id)
+}
+
+// -----------
+// | Getters |
+// -----------
+
+impl<'db, T: TransactionKind> StateTxn<'db, T> {
+    /// Get the name of the matching pool the given order is in, if it's been
+    /// allocated to one
+    pub fn get_pool_for_order(
+        &self,
+        order_id: &OrderIdentifier,
+    ) -> Result<Option<String>, StorageError> {
+        let pool_key = matching_pool_key(order_id);
+        self.inner().read(POOL_TABLE, &pool_key)
+    }
+
+    /// Whether or not a pool with the given name exists
+    pub fn pool_exists(&self, pool_name: &String) -> Result<bool, StorageError> {
+        let all_pools_key = all_pools_key();
+        let all_pools = self.read_set(POOL_TABLE, &all_pools_key)?;
+        Ok(all_pools.contains(pool_name))
+    }
+
+    /// Whether or not the matching pool is empty
+    pub fn pool_is_empty(&self, pool_name: &String) -> Result<bool, StorageError> {
```

Comment:
any reason this is `&String` and not `&str`? I know we require a non-reference type for DB access but you can see how other methods handle this with string keys at the storage layer; i.e. `&s.to_string()`. I feel this gives a cleaner interface as it accepts a wider range of types, and removes the need to do `to_string` in the caller. Plus would be nice to be consistent across the storage layer

Reply:
yeah, honestly was just nice to leverage type inference on the tx/cursor methods, but can use `&str` here